File: app/src/main/java/com/example/gymsessionsstopwatch_wearos/sessions/WorkoutSessionManager.java
package com.example.gymsessionsstopwatch_wearos.sessions;

import android.content.Context;

import com.example.gymsessionsstopwatch_wearos.local_files.LocalFileManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class WorkoutSessionManager {
    private List<WorkoutSession> sessions;
    private static Context appContext;
    private static final String DEFAULT_TIME = "0:00:00";

    public WorkoutSessionManager(Context context) {
        this.appContext = context.getApplicationContext();
        this.sessions = new ArrayList<>();
    }

    public WorkoutSession createNewSession() {
        WorkoutSession newSession = new WorkoutSession();
        sessions.add(newSession);
        updateSessionsFile();
        return newSession;
    }

    // Adds new block in the latest session
    public void addSessionBlock(WorkoutSession.SessionActivityType type, String startDate, long timedTime) {
        WorkoutSession session = sessions.get(sessions.size() - 1);
        if(session != null) {
            session.addSessionBlock(type, startDate, timedTime);
        }
        updateSessionsFile();
    }

    private void updateSessionsFile() {
        LocalFileManager.saveSessionsToFile(appContext, sessions);
    }

    public static List<WorkoutSession> getAllSessions() {
        List<WorkoutSession> loadedSessions = LocalFileManager.loadSessionFromFile(appContext);
        if (loadedSessions == null) {
            loadedSessions = new ArrayList<>();
        }
        return loadedSessions;
    }

    public static WorkoutSession.SessionBlock getLatestBlock() {
        List<WorkoutSession> loadedSessions = getAllSessions();

        if (loadedSessions.isEmpty()) {
            return null;
        }

        WorkoutSession latestSession = loadedSessions.get(loadedSessions.size() - 1);
        List<WorkoutSession.SessionBlock> blocks = latestSession.getSessionBlocks();

        if (blocks.isEmpty()) {
            return null;
        }

        return blocks.get(blocks.size() - 1);
    };

    public static String getLatestStopwatch() {
        WorkoutSession.SessionBlock latestBlock = getLatestBlock();
        if (latestBlock == null) {
            return DEFAULT_TIME;
        }

        long timedTime = latestBlock.getTimedTime();

        int hours = (int) (timedTime / 3600000);
        int minutes = (int) (timedTime % 3600000) / 60000;
        int seconds = (int) ((timedTime % 60000) / 1000);

        return String.format(Locale.getDefault(), "%02d:%02d:%02d", hours, minutes, seconds);
    }
}

File: app/src/main/java/com/example/gymsessionsstopwatch_wearos/sessions/WorkoutSessionManager.kt
package com.example.gymsessionsstopwatch_wearos.sessions

import android.content.Context
import com.example.gymsessionsstopwatch_wearos.local_files.LocalFileManager
import java.util.Locale

class WorkoutSessionManager(context: Context) {
    private var sessions: MutableList<WorkoutSession> = mutableListOf()
    private val appContext: Context = context.applicationContext

    init {
        sessions = mutableListOf()
    }

    fun createNewSession(): WorkoutSession {
        val newSession = WorkoutSession()
        sessions.add(newSession)
        updateSessionsFile()
        return newSession
    }

    fun addSessionBlock(type: WorkoutSession.SessionActivityType, startDate: String, timedTime: Long) {
        sessions.lastOrNull()?.let {
            it.addSessionBlock(type, startDate, timedTime)
            updateSessionsFile()
        }
    }

    private fun updateSessionsFile() {
        LocalFileManager.saveSessionsToFile(appContext, sessions)
    }

    companion object {
        private const val DEFAULT_TIME = "0:00:00"

        fun getAllSessions(context: Context): List<WorkoutSession> {
            return LocalFileManager.loadSessionFromFile(context) ?: listOf()
        }

        private fun getLatestBlock(context: Context): WorkoutSession.SessionBlock? {
            val loadedSessions = getAllSessions(context)
            val latestSession = loadedSessions.lastOrNull() ?: return null
            return latestSession.getSessionBlocks().lastOrNull()
        }

        fun getLatestStopwatch(context: Context): String {
            val latestBlock = getLatestBlock(context) ?: return DEFAULT_TIME
            val timedTime = latestBlock.timedTime
            val hours = (timedTime / 3600000).toInt()
            val minutes = ((timedTime % 3600000) / 60000).toInt()
            val seconds = ((timedTime % 60000) / 1000).toInt()
            return String.format(Locale.getDefault(), "%02d:%02d:%02d", hours, minutes, seconds)
        }
    }
}
